developers/weaviate/concepts/indexing/hnsw-indexes.mdx

```diff
@@ -0,0 +1,156 @@
+
+Weaviate's HNSW implementation is a very fast, memory efficient, approach to similarity search. The memory cache only stores the highest layer of the index instead of storing all of the data objects from the lowest layer. As a search moves from a higher layer to a lower one, HNSW only adds the data objects that are closest to the search query. This means HNSW uses a relatively small amount of memory compared to other search algorithms.
+
+The diagram demonstrates how the HNSW algorithm searches. The blue search vector in the top layer connects to a partial result in layer one. The objects in layer one lead HNSW to the result set in layer zero. In this example, HNSW makes three hops through the layers (the dotted blue lines) and skips the objects that are unrelated to the search query.
+
+## Search quality vs search speed
+
+HNSW parameters can be adjusted to adjust search quality against speed.
+
+The `ef` parameter is a critical setting for balancing the trade-off between search speed and quality.
+
+The `ef` parameter dictates the size of the dynamic list (the nearest neighbors list in the example above) that the HNSW algorithm uses during the search process. A higher `ef` value results in a more extensive search. A higher `ef` value improves accuracy but it can also slow down the query.
+
+A lower value for `ef` makes the search faster, but it might compromise on accuracy.
+
+The balance is crucial in scenarios where either speed or accuracy is a priority. For instance, in applications where rapid responses are critical, a lower `ef` might be preferable, even at the expense of some accuracy. In analytical or research contexts where precision is paramount, a higher `ef` might be better, despite the increased query time.
+
+### Configure ef
+
+You can set `ef` explicitly or allow Weaviate to set it dynamically. When `ef` is configured dynamically, Weaviate optimizes the balance between speed and recall based on real-time query requirements. If your query patterns are variable, consider setting dynamic `ef`.
+
+To enable dynamic `ef`, set `ef: -1` in your [collection configuration](/developers/weaviate/configuration/indexing-vector/hnsw-indexes#configure-an-hnsw-index). When dynamic `ef` is enabled, Weaviate adjusts the size of the ANN list based on the query response limit. The calculation also takes into account the values of `dynamicEfMin`, `dynamicEfMax`, and `dynamicEfFactor`.
+
+### Configure dynamic ef
+
+The `ef` parameter controls the size of the ANN list at query time. You can configure a specific list size or else let Weaviate configure the list dynamically. If you choose dynamic `ef`, Weaviate provides several options to control the size of the ANN list.
+
+The length of the list is determined by the query response limit that you set in your query. Weaviate uses the query limit as an anchor and modifies the size of ANN list according to the values you set for the `dynamicEf` parameters.
```

Comment:
> The length of the list is determined by the query response limit that you set in your query.

What happens if a limit is not specified? I think it would use the default limit.